tests/scanner/scanners/fw_rules_scanner_test.py

```diff
@@ -0,0 +1,437 @@
+            resource_id='project2', resource_type='project')
+        self.project3 = fre.resource_util.create_resource(
+            resource_id='project3', resource_type='project')
+        self.exception = fre.resource_util.create_resource(
+            resource_id='honeypot_exception', resource_type='project')
+        self.folder1 = fre.resource_util.create_resource(
+            resource_id='folder1', resource_type='folder')
+        self.folder2 = fre.resource_util.create_resource(
+            resource_id='test_instances', resource_type='folder')
+        self.folder3 = fre.resource_util.create_resource(
+            resource_id='folder3', resource_type='folder')
+        self.folder4 = fre.resource_util.create_resource(
+            resource_id='folder4', resource_type='folder')
+        self.org = fre.resource_util.create_resource(
+            resource_id='org', resource_type='organization')
+        self.project_resource_map = {
+            'test_project': self.project0,
+            'project1': self.project1,
+            'project2': self.project2,
+            'project3': self.project3,
+            'honeypot_exception': self.exception,
+        }
+        self.ancestry = {
+            self.project0: [self.folder1, self.org],
+            self.project1: [self.folder2, self.org],
+            self.project2: [self.folder4, self.folder3, self.org],
+            self.project3: [self.folder3, self.org],
+            self.exception: [self.folder3, self.org],
+        }
+
+    def test_get_output_filename(self):
+        """Test that the output filename of the scanner is correct.
+
+        Expected:
+            * Scanner output filename matches the format.
+        """
+        fake_utcnow_str = self.fake_utcnow.strftime(
+            self.scanner.OUTPUT_TIMESTAMP_FMT)
+
+        expected = self.scanner.SCANNER_OUTPUT_CSV_FMT.format(fake_utcnow_str)
+        actual = self.scanner._get_output_filename(self.fake_utcnow)
+        self.assertEquals(expected, actual)
+
+    @mock.patch(
+        'google.cloud.security.scanner.scanners.fw_rules_scanner.notifier',
+        autospec=True)
+    @mock.patch.object(
+        fw_rules_scanner.FwPolicyScanner,
+        '_upload_csv', autospec=True)
+    @mock.patch(
+        'google.cloud.security.scanner.scanners.fw_rules_scanner.os',
+        autospec=True)
+    @mock.patch(
+        'google.cloud.security.scanner.scanners.fw_rules_scanner.datetime',
+        autospec=True)
+    @mock.patch.object(
+        fw_rules_scanner.csv_writer,
+        'write_csv', autospec=True)
+    @mock.patch.object(
+        fw_rules_scanner.FwPolicyScanner,
+        '_output_results_to_db', autospec=True)
+    @mock.patch.object(
+        fw_rules_scanner.FwPolicyScanner,
+        '_flatten_violations')
+    # autospec on staticmethod will return noncallable mock
+    def test_output_results_local_no_email(
+        self, mock_flatten_violations, mock_output_results_to_db,
```

Comment:
This and the next heavily mocked tests are going to have to be rewritten every time the module under test changes. Please remove mocks or rewrite tests and code in a way that makes it transparent to changes in the underlying module. Nothing is really getting tested here in a way that would catch bugs, it will just catch the code changing.